language/move-prover/bytecode/src/compositional_analysis.rs

```diff
@@ -57,20 +57,21 @@ impl<'a> SummaryCache<'a> {
 }
 
 /// Trait that lifts an intraprocedural analysis into a bottom-up, compositional interprocedural
-/// analysis. The derived `summarize` function computes a postcondition for the procedure under
-/// analysis, then stores it in the summary cache for other procedures to use.
-// TODO: allow client analyses to customize this by providing a function that converts a
-// postcondition into a summary
+/// analysis.
 pub trait CompositionalAnalysis: DataflowAnalysis
 where
-    Self::State: 'static,
+    Self::State: AbstractDomain + 'static,
 {
-    fn summarize(
+    /// `analysis.summarize_abstract env init data abstrct` computes a postcondition for the
+    /// procedure `data` and then maps the postcondition to an element of abstract domain `T` by
+    /// applying the `abstrct` function. The result is stored in the summary cache of `data`.
+    fn summarize_abstract<T : AbstractDomain + 'static>(
         &self,
         func_env: &FunctionEnv<'_>,
         initial_state: Self::State,
         mut data: FunctionData,
-    ) -> FunctionData {
+        abstrct : fn(Self::State) -> T
```

Comment:
Why do we pass this as a function parameter? I would have expected this to be trait function which the user implements for this trait. In this case, you could also provide a default implementation for the case `Self::State == T`, I guess (but that is not that important).